util: add ordered option

BridgeAR opened the pull request:
The first commit adds the ordered option to `util.inspect()`. It makes sure object inspection output is ordered and reliable, no matter of the property insertion order or the order in the set / map.

The second commit improves the assertion diff output by using the ordered option. If requested, I'll move this PR out in a second one that relies on this.

Refs: #22763

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

targos commented:
I'd call it "sorted"

mscdex commented:
What does 'reliable' mean here exactly? To me that sounds like 'stable' which we don't have yet for sorting in V8 (but will soon as this behavior has been merged into V8 recently FWIW).

BridgeAR commented:
I have no strong opinion about either `ordered`, `sorted` or `stable` as name.

@mscdex I used `stable` in the commit description. If the sort is truly stable (as in the new V8 version) does not matter for this use case since all values are strings.

BridgeAR commented:
Small poll @nodejs/util for finding the best name 

:+1: ordered
:tada: sorted
:heart: stable

jdalton commented:
When I saw the issue title and began to review the PR I thought this was an option to preserve object property order _(I thought maybe the existing implementation may have messed that up and this was a back compat way to fix it)_ and not something dealing with sorting. I was surprised when I got to the code bit and it was for sorting properties.

BridgeAR commented:
I just renamed it to `sorted`. 

CI https://ci.nodejs.org/job/node-test-pull-request/17160/

jdalton commented on lib/util.js:
```diff
@@ -885,6 +887,13 @@ function formatRaw(ctx, value, recurseTimes) {
+    if (ctx.sorted === true) {
+      output = output.sort();
+      return reduceToSingleString(ctx, output.sort(), base, braces);
+    }
```
☝️ This is sorting things twice. I believe the first one `output = output.sort()` isn't needed.

jdalton commented on lib/util.js:
```diff
@@ -885,6 +887,13 @@ function formatRaw(ctx, value, recurseTimes) {
+    }
+    return reduceToSingleString(ctx, output.sort(ctx.sorted), base, braces);
+  }
   return reduceToSingleString(ctx, output, base, braces);
```
👆 To dry out the condition above it could be something like:
```js
const comparator = ctx.sorted === true ? undefined : ctx.sorted
return reduceToSingleString(ctx, output.sort(comparator), base, braces)
```

BridgeAR commented:
Comments addressed :)

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/17161/

addaleax commented on doc/api/util.md:
```diff
@@ -426,6 +429,10 @@ changes:
+  * `sorted` {boolean|Function} If specified, the output for all object types
+    besides arrays and typed arrays is going to be stable. If set to `true` the
+    default sort is going to be used. If set to a function, it is used as
+    [compare function][].
```
`as a [compare function][]`?

addaleax commented on doc/api/util.md:
```diff
@@ -426,6 +429,10 @@ changes:
     characters, no matter the `breakLength` size. For more information, see the
     example below. **Default:** `true`.
+  * `sorted` {boolean|Function} If specified, the output for all object types
+    besides arrays and typed arrays is going to be stable. If set to `true` the
```
I’m not sure it’s obvious what `stable` means? Maybe just explain what happens, i.e. that entries will be sorted before being printed?

BridgeAR commented:
PTAL. I just rebased due to conflicts and addressed the comments. I also added an example and made sure that non-index properties on arrays and typed arrays will also be sorted. The changed part is in the `fixup: util` commit. 

CI https://ci.nodejs.org/job/node-test-pull-request/17225/

vsemozhetbyt commented on doc/api/util.md:
````diff
@@ -535,6 +542,34 @@ console.log(inspect(weakSet, { showHidden: true }));
 // WeakSet { { a: 1 }, { b: 2 } }
 ```
 
+The `sorted` option makes sure the output is stable, no matter of the properties
````
It seems the word `stable` is a bit confusing here. Compare this sentence with the [wiki definition](https://en.wikipedia.org/wiki/Sorting_algorithm#Stability) (referenced from the MDN `Array.prototype.sort()` page):

> Stable sort algorithms sort identical elements in the same order that they appear in the input.

I.e. the traditional meaning of the `stable` sorting is opposite to `no matter of the properties insertion order`.

BridgeAR replied:
True, I think I found a better wording.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/17283/
